## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [343]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [344]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [345]:
# Use the describe function to gather some basic statistics
df.describe

<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [346]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info

<bound method DataFrame.info of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  

In [347]:
# What three item categories had the most entries?
top_three_categories = df['category'].value_counts().head(3)
top_three_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [348]:
# For the category with the most entries, which subcategory had the most entries?

most_common_category = df['category'].value_counts().idxmax()

filtered_df = df[df['category'] == most_common_category]

most_common_subcategory = filtered_df['subcategory'].value_counts().idxmax()
most_common_category

'consumables'

In [349]:
# Which five clients had the most entries in the data?
top_five_clients = df['client_id'].value_counts().head(5)
top_five_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [350]:
# Store the client ids of those top 5 clients in a list.
top_five_clients.tolist()

[220, 211, 209, 207, 207]

In [351]:
# How many total units (the qty column) did the client with the most entries order order?
most_entries_client = df['client_id'].value_counts().idxmax()
total_units_ordered = df[df['client_id'] == most_entries_client]['qty'].sum()
total_units_ordered

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [352]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal'],
      dtype='object')

In [353]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping(unit_weight):
    if unit_weight > 50:
        return unit_weight*7
    else:
        return unit_weight*10

df['Shipping Price'] = df['unit_weight'].apply(calculate_shipping)
df



first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  Shipping Price  
0                1  115164.00            75.0  
1                0     523.95            14.9  
2                6     527.28            16.8  
3                3    1056.18            12.3  
4                1    3902.00           464.3  
...            ...        ...             ...  
54634            8    2743.29            22.5  
54635            0    9709.73           117.0  
54636            6   31188.50            41.6  
54637            0     165.76           180.4  
54638            5    3

In [354]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['Total Price'] = ((df['subtotal'] + df['Shipping Price']) *1.0925)

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number   subtotal  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       Shipping Price    Total Price  
0                75.0  125898.607500  
1                14.9     588.693625  
2                16.8     594.407400  
3                12.3    1167.314400  
4               464.3    4770.182750  
...               ...            ...  
54634            22.5    3021.625575  
54635           117.0   10735.702525  
54636            41.6   34118.884250  
54637           1

In [355]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['Item Cost'] = (df['unit_cost'] * df['qty']) + df['Shipping Price'] 
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   subtotal  Shipping Price  \
0             7.50   105            1  115164.00            75.0   
1             1.49    21            0     523.95            14.9   
2             1.68    39            6     527.28            16.8   
3             1.23    29            3    1056.18            12.3   
4            46.43    20            1    3902.00           464.3   
...            ...   ...          ...        ...             ...   
54634         2.25    33            8    2743.29            22.5   
54635        11.70    47            0    9709.73           117.0   
54636         4.16   475            6   31188.50            41.6   
54637        18.04   112            0     165.76           180.4   
54638         2.07  1031            5    3103.31            20.7   

         Total Price  Item Cost  
0      125898.607500   80159.55  
1         588.693625     331.79  
2         594.407400     323.34  
3        1167.314400     732.95  
4        4770.182750    2627.70  
...              ...        ...  
54634    3021.625575    1725.30  
54635   10735.702525    8363.62  
54636   34118.884250   27263.85  
54637     378.179800     317.04  
54638    

In [356]:
# Create a column for the profit of each line using line cost and line price
df['Profit'] = df['Total Price'] - df['Item Cost']
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  Shipping Price    Total Price  Item Cost  \
0                1  115164.00            75.0  125898.607500   80159.55   
1                0     523.95            14.9     588.693625     331.79   
2                6     527.28            16.8     594.407400     323.34   
3                3    1056.18            12.3    1167.314400     732.95   
4                1    3902.00           464.3    4770.182750    2627.70   
...            ...        ...             ...            ...        ...   
54634            8    2743.29            22.5    3021.625575    1725.30   
54635            0    9709.73           117.0   10735.702525    8363.62   
54636            6   31188.50            41.6   34118.884250   27263.85   
54637            0     165.76           180.4     378.179800     317.04   
54638            5    3103.31            20.7    3412.980925    1680.61   

             Profit  
0      45739.057500  
1        256.903625  
2        271.067400  
3        434.364400  
4       2142.482750  
...             ...  
54634   1296.325575  
54635   2372.082525  
54636   6855.034250  
54637     61.139800  
54638   1732.370925  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [357]:
# Check your work using the totals above
# Function to calculate total price for a specific Order ID
def calculate_total(order_id):
    order_data = df[df['order_id'] == order_id]
    return order_data['Total Price'].sum()

# Check totals for each Order ID
total_2742071 = calculate_total(2742071)
total_2173913 = calculate_total(2173913)
total_6128929 = calculate_total(6128929)

print(f'Total for Order ID 2742071: {total_2742071}')
print(f'Total for Order ID 2173913: {total_2173913}')
print(f'Total for Order ID 6128929: {total_6128929}')

Total for Order ID 2742071: 139844.51202499997
Total for Order ID 2173913: 119838.35704999999
Total for Order ID 6128929: 752748.7454750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [358]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients = df.groupby('client_id')['qty'].sum().nlargest(5).index

total_spending = df[df['client_id'].isin(top_clients)].groupby('client_id')['Total Price'].sum()

summary_df = total_spending.reset_index()
summary_df.columns = ['client_id', 'Total__Price']


print(summary_df)



   client_id  Total__Price
0      30143  2.351848e+07
1      77354  4.592791e+07
2      78965  6.500970e+08
3      85931  2.263611e+08
4      86721  1.109862e+08


In [359]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_df = df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('Shipping Price', 'sum'),
    total_revenue=('Total Price', 'sum'),
    total_profit=('Profit', 'sum')
).reset_index()

# Sort by total revenue and get the top 5 clients
top_clients_df = summary_df.sort_values(by='Total Price', ascending=False).head(5)

# Display the summary DataFrame
print(top_clients_df)

KeyError: 'Total Price'

In [360]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['Total Price', 'Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return x / 1_000_000

# Apply the currency_format_millions function to only the money columns. 
df[money_columns] = df[money_columns].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
df.rename(columns={
    'Total Price': 'Total Price (in millions)',
    'Total Profit': 'Total Profit (in millions)'
}, inplace=True)


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  Shipping Price  Total Price (in millions)  \
0                1  115164.00            75.0                   0.125899   
1                0     523.95            14.9                   0.000589   
2                6     527.28            16.8                   0.000594   
3                3    1056.18            12.3                   0.001167   
4                1    3902.00           464.3                   0.004770   
...            ...        ...             ...                        ...   
54634            8    2743.29            22.5                   0.003022   
54635            0    9709.73           117.0                   0.010736   
54636            6   31188.50            41.6                   0.034119   
54637            0     165.76           180.4                   0.000378   
54638            5    3103.31            20.7                   0.003413   

       Item Cost    Profit  
0       80159.55  0.045739  
1         331.79  0.000257  
2         323.34  0.000271  
3         732.95  0.000434  
4        2627.70  0.002142  
...          ...       ...  
54634    1725.30  0.001296  
54635    8363.62  0.002372  
54636   27263.85  0.006855  
54637     317.04  0.0

In [363]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df.sort_values("Total Price (in millions)", ascending=False)

first      last                                    job  \
21801     Julie  Anderson                    Exhibition designer   
15112    Dennis   Hawkins    Social research officer, government   
23929   Michele  Mcdonald  Emergency planning/management officer   
25972       Amy   Stewart              Psychologist, educational   
11740    Lauren     Jones                              Architect   
...         ...       ...                                    ...   
22529   Matthew    Orozco             Research scientist (maths)   
37936  Jennifer     Lewis                Agricultural consultant   
29972    Dennis   Hawkins    Social research officer, government   
21090    Daniel    Bowman                          Archaeologist   
38399     David     Green                      Personnel officer   

                       phone                            email  client_id  \
21801             5933099462    anderson_better_4704@bell.biz      78965   
15112  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
23929     435-931-4939x95137    mcdonaldm317@hess-jackson.com      38473   
25972     979.793.4884x57205           stewartamy@francis.com      86721   
11740           785-820-6323   jones.security.7283@gibson.com      84344   
...                      ...                              ...        ...   
22529   +1-259-213-7449x4333           orozcomatthew@carr.com      67407   
37936             9813732574  lewis.jennifer@becker-ford.info      62566   
29972  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
21090   001-716-757-9724x862        danielbowman@thompson.com      72892   
38399     527.445.6304x24709    dgreen4039@sutton-leonard.org      57283   

       order_id  order_date  order_week  order_year  ... unit_price unit_cost  \
21801   3833568  2023-04-29          17        2023  ...     147.66    100.11   
15112   7263857  2023-03-05           9        2023  ...     323.42    183.84   
23929   4594154  2023-03-02           9        2023  ...     674.94    568.06   
25972   6057452  2023-03-21          12        2023  ...     100.40     73.05   
11740   7071092  2023-03-13          11        2023  ...     211.44    171.21   
...         ...         ...         ...         ...  ...        ...       ...   
22529   5546401  2023-05-18          20        2023  ...       0.13      0.12   
37936   5131256  2023-05-17          20        2023  ...       0.13      0.12   
29972   7438979  2023-03-27          13        2023  ...       0.13      0.12   
21090   6514417  2023-05-17          20        2023  ...      96.86     74.98   
38399   6619389  2023-05-22          21        2023  ...      58.39     49.69   

      unit_weight      qty  line_number      subtotal  Shipping Price  \
21801        0.09  3958244            1  5.844743e+08             0.9   
15112        2.21   637318            3  2.061214e+08            22.1   
23929        2.25   219645            5  1.482472e+08            22.5   
25972        9.22   977121            3  9.810295e+07            92.2   
11740        1.28   389960            4  8.245314e+07            12.8   
...           ...      ...          ...           ...             ...   
22529        0.28        8            0  1.040000e+00             2.8   
37936        0.28        6            2  7.800000e-01             2.8   
29972        0.28        1            6  1.300000e-01             2.8   
21090        0.24        0            6  0.000000e+00             2.4   
38399        0.18        0            0  0.000000e+00             1.8   

       Total Price (in millions)     Item Cost        Profit  
21801                 638.538184  3.962598e+08  2.422784e+02  
15112                 225.187640  1.171646e+08  1.080231e+02  
23929                 161.960087  1.247716e+08  3.718853e+01  
25972                 107.177572  7.137878e+07  3.579879e+01  
11740                  90.080072  6.676506e+07  2.331501e+01  
...                          ...           ...      